# Sentiment Analysis Using Emojis Project

SuiLin's notes, February 27, 2016


### Data Exploration on Labels

In this notebook we explore the connections between the emoji based labels and the human ones, based on a random sample of 500 tweets. the objective is to establish whether the emojis provide an accurate enough labeling, for the two sentiments analyzed (happy, unhappy).

In [4]:
import numpy as np

In [2]:
f = open("d:\\data\\tweetvalidate.tsv", "r") # load the validated data (human labelled tweets), "r" means open the file in read-only mode.
lines = f.readlines() # Read all the lines in the file.
f.close()

In [5]:
N = len(lines) - 2 # The first 2 lines are the headers.

In [6]:
VD = np.zeros(N) # Validated Data array with size N. Initialize to all the values in the array to 0.

In [22]:
c = 0

In [23]:
for line in lines[2:]: # skip the first 2 lines (headers). For each line in the lines.
    # Split into columns in the tsv file and assign column #1 to column #3 to temp. Column #0, which is the ID is skipped.
    temp = line.split("\t")[1:4] 
    
    for i in xrange(3): # Iterate through the loop for i=0, i=1, i=2.
        if temp[i] == "p" or temp[i] == "P":
            temp[i] = 1
        elif temp[i] == "n" or temp[i] == "N":
            temp[i] = -1
        else:
            temp[i] = 0
    
    # Sum up the values in temp and then sign() will return either (-1, 0, 1) depending on the labels.
    VD[c] = np.sign(sum(temp)) 
    c += 1 # Go to the next line in the tsv file.

In [25]:
f = open("d:\\data\\tweetvalidatekey.csv", "r") # load the true data (emoji labelled tweets
lines = f.readlines()
f.close()
# TODO: Might want to reset N to the lines in csv?
#N = len(lines) -1 # Skip the first line since it is the header.

In [34]:
TD = np.ones(N) # True Data. Create an array with N elements and initailize the elements to 1.

In [52]:
c = 0

In [53]:
for line in lines[1:N]: # skip the first line (headers)
    line = line.split(",") # line = [tweet_id, sentiment, tweets]
    label = line[1]  # Assign sentiment to label.
    if label == "Negative": TD[c] = -1 # Set True Data to -1 if the label is "Negative".
    c += 1  # Go to the following line.

In [56]:
CM = np.zeros((3,2)) # Confusion Matrix. Create a 2-dimensional array with 3 rows and 2 columns and initialize the values to 0.

In [57]:
VL = [1, -1, 0] # Validated Labels
TL = [1, -1] # True Labels

In [2]:
for z in xrange(N): # xrange is from 0 to N-1. N is the number of tweets in the files.
    for i in xrange(3): # Iterate through validated labels.
        for j in xrange(2): # Iterate through true labels.
            if (VD[z] == VL[i]) and (TD[z] == TL[j]): CM[i,j] += 1 # increase the values in confusion matrix

print CM # 103 tweets are emoji labelled and validated to be positive.  
         # 205 tweets are emoji labelled as negative but validated as positive.

NameError: name 'N' is not defined

In [68]:
# Similarity Index (ignoring 3rd row). Take sum of correctly labelled postive and negative tweets and divide by all the tweets
# validated to as positive or negative (excludes neutral tweets)
print SI
SI = (CM[0,0] + CM[1,1]) / (N - sum(CM[2,:])) 

0.539956803456


In [69]:
def JS(CM, C): # Jaccard Similarity of class C TODO: Ask why exclude other intersection
#BP = both positive; BN = both negative; E= emoji; H=human
    N = sum(sum(CM))
    
    if C == 0: #BP divided by BP + ENHP + EPHN
        return CM[0,0] / (N - CM[1,1])
    else: #BN divided by BN + ENHP + EPHN
        return CM[1,1] / (N - CM[0,0])

In [72]:
J = [JS(CM,0), JS(CM,1)]
print J

[0.29178470254957506, 0.37027707808564231]


### Conclusions

There is some overlap between the human-labeled tweets and the emoji labels (54%), yet there is a large number of false positives (if we take the emoji data to be correct). This is reflected by the Jaccard Similarity on the "Positive" class (0), which is around 29%. The number of neutral tweets, although small, is not negligible (~ 7.5%). Perhaps the use of a third class in the emoji classification would improve the overall similarity.